CATS/JudgeDB.pm

```diff
@@ -633,6 +633,27 @@ sub insert_req_details {
     1;
 }
 
+sub get_input_hash {
+    my ($problem_id, $test_rank) = @_;
+
+    $dbh->selectrow_array(q~
+        SELECT in_file_hash FROM tests WHERE problem_id = ? AND rank = ?~, undef,
+    $problem_id, $test_rank);
+}
+
+sub save_input_hash {
+    my ($problem_id, $test_rank, $hash) = @_;
+
+    eval {
+        $dbh->do(q~
+        UPDATE tests SET in_file_hash = ? WHERE problem_id = ? AND rank = ?~, undef,
```

Comment:
Add `AND (in_file_hash IS NULL OR in_file_hash = ?)` condition to avoid race.
Check update count (return value of `do` ) to make sure that condition actually worked, otherwise report error.

Reply:
Shouldn't there be the same check in other similar functions(save_input_test_data, ..)? 'report error' is just warning, the problem installation should be continued, right?